api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java

```diff
@@ -492,7 +507,85 @@ private void attemptFirecloudWorkspaceCreation(FirecloudWorkspaceId workspaceId)
   @Override
   public ResponseEntity<CloneWorkspaceResponse> cloneWorkspace(String workspaceNamespace,
       String workspaceId, CloneWorkspaceRequest body) {
-    return ResponseEntity.status(HttpStatus.NOT_IMPLEMENTED).build();
+    Workspace workspace = body.getWorkspace();
+    if (Strings.isNullOrEmpty(workspace.getNamespace())) {
+      throw new BadRequestException("missing required field 'workspace.namespace'");
+    } else if (Strings.isNullOrEmpty(workspace.getName())) {
+      throw new BadRequestException("missing required field 'workspace.name'");
+    } else if (workspace.getResearchPurpose() == null) {
+      throw new BadRequestException("missing required field 'workspace.researchPurpose'");
+    }
+    User user = userProvider.get();
+    if (user == null) {
+      // You won't be able to create workspaces prior to creating a user record once our
+      // registration flow is done, so this should never happen.
+      throw new BadRequestException("User is not initialized yet; please register");
+    }
+    if (workspaceService.getByName(workspace.getNamespace(), workspace.getName()) != null) {
+      throw new ConflictException(String.format(
+          "Workspace %s/%s already exists",
+          workspace.getNamespace(), workspace.getName()));
+    }
+
+    checkWorkspaceReadAccess(workspaceNamespace, workspaceId);
+    org.pmiops.workbench.db.model.Workspace fromWorkspace = workspaceService.getRequired(
+        workspaceNamespace, workspaceId);
+    if (fromWorkspace == null) {
+      throw new NotFoundException(String.format(
+          "Workspace %s/%s not found", workspaceNamespace, workspaceId));
+    }
+
+    FirecloudWorkspaceId fcWorkspaceId = generateFirecloudWorkspaceId(workspace.getNamespace(),
+        workspace.getName());
+    fireCloudService.cloneWorkspace(workspaceNamespace, workspaceId,
+        fcWorkspaceId.getWorkspaceNamespace(), fcWorkspaceId.getWorkspaceName());
+
+    // TODO(calbach): Copy GCS scratch files.
```

Comment:
It sounds like we may not need to do this -- FC will clone notebooks for us.

Reply:
Updated TODO since the timelines are unclear; figuring this out now.